### ISSUE TITLE
Should the browser allow scrolling "out of bounds" in resizes-visual or overlays-content mode?

### ISSUE THREAD

**Issue by theres-waldo** [bramus/viewport-resize-behavior#1]
We're stating to think about what would be involved in implementing this in Firefox, and I'd like to ask for a clarification:

In `resizes-visual` or `overlays-content` mode, if the input element that triggered the keyboard is near the bottom of the page (note: I'm not talking about `position: fixed` or anything, just regular in-flow scrolled content that's near the bottom of a scrollable page), that input element will, at least in a naive implementation, end up **behind** the keyboard.

Is it the intention that in such cases, the browser allows scrolling "out of bounds" to scroll that input element into view above the keyboard (such that some of the area behind the keyboard would now be an out-of-bounds region past the scroll-height of the page -- but maybe that's fine, since it's hidden by the keyboard anyways)?

Or is it the responsibility of page authors to ensure this doesn't happen, and if it does then too bad, you end up typing into an input field and you can't see what you're typing?

cc @hiikezoe

**Comment by bramus** [comment]
Through testing I found that is only in `overlays-content` mode that Chrome (on Android) does not try to scroll the focused element into view. In that case, the input can end up behind the keyboard indeed. Adding @bokand to confirm that this is the implemented behavior.

It is documented that in `overlays-content` mode, the contents of the website can be obscured by the keyboard and that the author must use [the exposed environment variables](https://w3c.github.io/virtual-keyboard/#virtual-keyboard-visibility-change-css-environment-variables) should they want to keep elements outside of the covered area.

Note that this is not an official CSSWG / Google / Chrome repo, but merely a repo that includes [a bunch of demo pages](https://viewport-resize-behavior.netlify.app/) to test the various modes. If this extra clarification about autoscroll is missing in the spec, then it should be added to the spec by filing an issue in [its proper repository](https://github.com/w3c/virtual-keyboard/).

**Comment by bokand** [comment]
That's right - as I understand it, `overlays-content` is a signal to the UA that the page author will take manual steps to react to the virtual keyboard to ensure the UI is adjusted appropriately and so the UA shouldn't attempt to react _at all_ to the virtual keyboard (other than giving the page information about the keyboard geometry). [This example](https://w3c.github.io/virtual-keyboard/#example-repositioning-content-using-css-with-overlayscontent) is instructive in how the authors of this API imagine UIs to respond to the keyboard.

`resizes-visual` (the default in Chrome and Safari) tells the UA that the virtual keyboard should resize the visual viewport but not the ICB or fixed viewport. i.e. don't affect the page's content but still allow scrolling to what would otherwise be unreachable areas. In this case, the "scroll focused input into view" steps Chrome takes will pan the visual viewport such that it becomes visible to the user. This is spec'd implicitly via the [Scroll a target into view steps](https://drafts.csswg.org/cssom-view/#scroll-a-target-into-view), when scrolling a viewport we use the [Perform a scroll of a viewport steps](https://drafts.csswg.org/cssom-view/#viewport-perform-a-scroll) which includes adjusting the visual viewport if it's insetted by the keyboard (or zoomed in).

**Comment by theres-waldo** [comment]
Thanks for the clarifications!

I think I had the wrong mental model for `resizes-visual`. I imagined it as being similar to `overlays-content` in that the browser would still render content underneath the keyboard, it would just adjust the `visualViewport` so that it reports the area above the keyboard; but I see now that it's more like the browser resizes its rendering area to be the area above the keyboard but without reflowing the page into that smaller area.

As such, `resizes-visual` is not affected by the issue I was concerned about (not being able to scroll the input element into view). `overlays-content` _is_ affected, but we're making the deliberate choice to make that the page's problem, which is fine.

> If this extra clarification about autoscroll is missing in the spec, then it should be added to the spec by filing an issue in [its proper repository](https://github.com/w3c/virtual-keyboard/).

What is the relationship between `interactive-widgets`  and the VirtualKeyboard API?

As an implementer, should we view one of these as a dependency of the other? Or as two alternative proposals? Or two complementary proposals which can be implemented in either order?

**Comment by bokand** [comment]
> What is the relationship between interactive-widgets and the VirtualKeyboard API?
>
> As an implementer, should we view one of these as a dependency of the other? Or as two alternative proposals? Or two complementary proposals which can be implemented in either order?

`interactive-widgets` is an API that enables a page to choose a virtual keyboard mode. The VirtualKeyboard API introduces one such mode (with some other API surfaces) but provides its own opt-in mechanism (`navigator.virtualKeyboard.overlaysContent = true;`)

As such, I think there's no dependency on each other.

You could implement `interactive-widgets` without the VirtualKeyboard API by respecting only the `resizes-visual` or `resizes-content` values.

You could implement the VirtualKeyboard API without `interactive-widgets` (that was the ordering in Chrome) since it provides its own JS opt-in.

The two were separate projects in Chrome. We added `overlays-content` to `interactive-widgets` since it seemed like a good fit and it was a common point of feedback on the VK API that it should support a declarative opt-in.

**Comment by theres-waldo** [comment]
Thanks, that makes sense!

Just to double check, since @bramus indicated that https://github.com/w3c/virtual-keyboard/ would be the place to file a spec issue about clarifying the behaviour in various virtual keyboard modes -- organizationally, is `interactive-widgets` going to be proposed as a modification/addition to the Virtual Keyboard spec?

**Comment by bokand** [comment]
I think anything related to the semantics of `overlaysContents` does belong in w3c/virtual-keyboard but IMHO issues related to the other values and meta tag do not. AFAICT virtual-keyboard was specific to the overlaysContent behavior and I'm not sure how active that spec/repo still is.

I spec'd `interactive-widgets` in https://drafts.csswg.org/css-viewport/#interactive-widget-section so I think CSSWG is probably the right place for more general questions.

**Comment by bramus** [comment]
> Just to double check, since @bramus indicated that https://github.com/w3c/virtual-keyboard/ would be the place to file a spec issue

Apologies, I should have linked to the `css-viewport` spec instead. @bokand was already kind enough to clarify things.
